File: app/src/main/java/com/ramez/shopp/activities/PayUsingCardActivity.java
//package com.ramez.shopp.activities;
//
//import static com.ramez.shopp.classes.JavaApplication1.generateRandomPassword;
//
//import android.annotation.SuppressLint;
//import android.content.Intent;
//import android.os.Bundle;
//import android.util.Log;
//
//import com.google.gson.Gson;
//import com.ramez.shopp.R;
//import com.ramez.shopp.Utils.DateHandler;
//import com.ramez.shopp.Utils.NumberHandler;
//import com.ramez.shopp.classes.CardInformation;
//import com.ramez.shopp.classes.Constants;
//import com.ramez.shopp.databinding.ActivityPayUsingCardBinding;
//
//import org.jetbrains.annotations.Nullable;
//
//import java.util.Base64;
//
//import javax.crypto.Cipher;
//import javax.crypto.spec.IvParameterSpec;
//import javax.crypto.spec.SecretKeySpec;
//
// public class PayUsingCardActivity extends ActivityBase {
//
//        ActivityPayUsingCardBinding binding;
//    String currentMonth, currentYear;
//
//
//    protected void onCreate(@Nullable Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//        binding = ActivityPayUsingCardBinding.inflate(this.getLayoutInflater());
//        this.setContentView(binding.getRoot());
//        currentMonth = DateHandler.GetMonthOnlyString();
//        currentYear = DateHandler.GetYearOnlyString();
//        Log.i("tag", "Log currentMonth " + currentMonth);
//        Log.i("tag", "Log currentYear " + currentYear);
//        binding.btnDone.setOnClickListener(v -> {
//
//
//            try {
//                String monthStr = NumberHandler.arabicToDecimal(binding.sprmonth.getText().toString());
//                String yearStr = NumberHandler.arabicToDecimal(binding.spryear.getText().toString());
//                String cardNumberStr = NumberHandler.arabicToDecimal(binding.edtCardNumber.getText().toString());
//                String cvvStr = NumberHandler.arabicToDecimal(binding.edtCvvNumber.getText().toString());
//
//                boolean hasError = false;
//
//
//                if (monthStr.isEmpty()) {
//                    binding.sprmonth.setError(getString(R.string.invalid_input));
//                    hasError = true;
//                }
//
//                if (yearStr.isEmpty()) {
//                    binding.spryear.setError(getString(R.string.invalid_input));
//                    hasError = true;
//                }
//
//                if (cardNumberStr.isEmpty() || cardNumberStr.length() < 14) {
//                    binding.edtCardNumber.setError(getString(R.string.invalid_input));
//                    hasError = true;
//                }
//
//                if (cvvStr.isEmpty() || cvvStr.length() < 3) {
//                    binding.edtCvvNumber.setError(getString(R.string.invalid_input));
//                    hasError = true;
//                }
//
//                if (Integer.parseInt(yearStr) < Integer.parseInt(currentYear)) {
//                    binding.spryear.setError(getString(R.string.invalid_input));
//                    hasError = true;
//                }
//
//                if (Integer.parseInt(monthStr) > 12) {
//                    binding.sprmonth.setError(getString(R.string.invalid_input));
//                    hasError = true;
//                }
//
//                Log.i("tag", "Log yearStr.toInt() "+Integer.parseInt(yearStr));
//
//                if (hasError)
//                    return;
//
//                CardInformation c = new CardInformation();
//                c.number = cardNumberStr;
//                c.month = monthStr;
//                c.year = yearStr;
//                c.securityCode = cvvStr;
//                sendData(c);
//
//            } catch (Exception e) {
//                e.printStackTrace();
//
//            }
//
//        });
//
//
//    }
//
//    private void sendData(CardInformation cardModel) {
//        Gson gson = new Gson();
//        String json = gson.toJson(cardModel);
//        String EncryptedString = null;
//        try {
//           // EncryptedString = Encrypt("{\"number\":\"0123456789123456\",\"month\":\"12\",\"year\":\"22\",\"securityCode\":\"123\"}");
//            EncryptedString = Encrypt(json);
//            String var5 = "Log [EncryptedString]: " + EncryptedString;
//            System.out.println(var5);
//            var5 = "Log  [json]: " + json;
//            System.out.println(var5);
//            Intent intent = new Intent();
//            intent.putExtra(Constants.PAY_TOKEN, EncryptedString);
//            setResult(-1, intent);
//            finish();
//        } catch (Exception e) {
//            e.printStackTrace();
//        }
//
//    }
//
//
//    public static String Encrypt(String text)
//            throws Exception {
//        String ivkey = "CDDAMQOTMYIAZEPQ";
////        String key = "h3tTWAYJ55EGWMgZFs5gW5mquCIsgLhE";
//          String key = generateRandomPassword(32);
//        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
//        byte[] keyBytes = new byte[32];
//        byte[] b = key.getBytes("UTF-8");
//        int len = b.length;
//        if (len > keyBytes.length) len = keyBytes.length;
//        System.arraycopy(b, 0, keyBytes, 0, len);
//
//        byte[] ivBytes = new byte[16];
//        b = ivkey.getBytes("UTF-8");
//        len = b.length;
//        if (len > ivBytes.length) len = ivBytes.length;
//        System.arraycopy(b, 0, ivBytes, 0, len);
//
//        SecretKeySpec keySpec = new SecretKeySpec(keyBytes, "AES");
//
//        IvParameterSpec ivSpec = new IvParameterSpec(ivBytes);
//        cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec);
//
//        byte[] results = cipher.doFinal(text.getBytes("UTF-8"));
//        @SuppressLint({"NewApi", "LocalSuppress"})
//        String EncryptedString = Base64.getEncoder().encodeToString(results);
//        return EncryptedString.substring(0, EncryptedString.length() / 2)
//                + key
//                + EncryptedString.substring(EncryptedString.length() / 2);
//    }
//}

File: app/src/main/java/com/ramez/shopp/activities/PayUsingCardActivity.kt
package com.ramez.shopp.activities

import android.content.Intent
import android.os.Bundle
import android.util.Base64
import android.util.Log
import android.view.View
import com.google.gson.Gson
import com.ramez.shopp.Models.CreditCardModel
import com.ramez.shopp.R
import com.ramez.shopp.Utils.DateHandler
import com.ramez.shopp.Utils.NumberHandler
import com.ramez.shopp.classes.Constants
import com.ramez.shopp.databinding.ActivityPayUsingCardBinding
import java.util.*
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec


class PayUsingCardActivity : ActivityBase() {
    lateinit var binding: ActivityPayUsingCardBinding


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPayUsingCardBinding.inflate(layoutInflater)
        val view: View = binding.root
        setContentView(view)

        var currentMonth: String? = null
        var currentYear: String? = null

        currentMonth= DateHandler.GetMonthOnlyString()
        currentYear=DateHandler.GetYearOnlyString()

        Log.i("tag","Log currentMonth ${currentMonth}")
        Log.i("tag","Log currentYear ${currentYear}")
//        currentMonth=  if (currentMonth.startsWith("0")) currentMonth.replaceFirst(
//            "0",
//            ""
//        ) else currentMonth
        binding.btnDone.setOnClickListener {
            try {
                var monthStr = NumberHandler.arabicToDecimal(binding.sprmonth.text.toString())
                val yearStr = NumberHandler.arabicToDecimal(binding.spryear.text.toString())
                val cardNumberStr = NumberHandler.arabicToDecimal(binding.edtCardNumber.text.toString())
                val cvvStr = NumberHandler.arabicToDecimal(binding.edtCvvNumber.text.toString())

                var hasError = false

//                monthStr=  if (monthStr.startsWith("0")) monthStr.replaceFirst(
//                    "0",
//                    ""
//                ) else monthStr

                if (monthStr.isEmpty()) {
                    binding.sprmonth.error = getString(R.string.invalid_input)
                    hasError = true
                }

                if (yearStr.isEmpty()) {
                    binding.spryear.error = getString(R.string.invalid_input)
                    hasError = true
                }

                if (cardNumberStr.isEmpty() || cardNumberStr.length <14) {
                    binding.edtCardNumber.error = getString(R.string.invalid_input)
                    hasError = true
                }

                if (cvvStr.isEmpty() ||cvvStr.length <3) {
                    binding.edtCvvNumber.error = getString(R.string.invalid_input)
                    hasError = true
                }

                if (yearStr.toInt() < currentYear.toInt()) {
                    binding.spryear.error = getString(R.string.invalid_input)
                    hasError = true
                }

                if (monthStr.toInt() > 12) {
                    binding.sprmonth.error = getString(R.string.invalid_input)
                    hasError = true
                }

                Log.i("tag","Log yearStr.toInt() ${yearStr.toInt()}")

                if (hasError)
                    return@setOnClickListener

                val cardModel = CreditCardModel()
                cardModel.number = cardNumberStr
                cardModel.month = monthStr
                cardModel.year = yearStr
                cardModel.securityCode = cvvStr

                sendData(cardModel)



            } catch (e: Exception) {
                e.printStackTrace()

            }

        }


    }

    private fun sendData(cardModel: CreditCardModel) {
        val gson = Gson()
        val json = gson.toJson(cardModel)
        val EncryptedString: String = Encrypt(json)
        println("Log [EncryptedString]:$EncryptedString")
        println("Log  [json]:$json")
        val intent = Intent()
        intent.putExtra(Constants.PAY_TOKEN,EncryptedString)
        setResult(RESULT_OK, intent)
        finish()


    }

    @Throws(java.lang.Exception::class)
    fun Encrypt(text: String?): String {
        val ivkey = "CDDAMQOTMYIAZEPQ"
        val key: String? = generateRandomPassword(32)
        val cipher: Cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        val keyBytes = ByteArray(32)
        var b = key?.toByteArray(charset("UTF-8"))
        var len = b?.size?:0
        if (len > keyBytes.size) len = keyBytes.size
        System.arraycopy(b, 0, keyBytes, 0, len)
        val ivBytes = ByteArray(16)
        b = ivkey.toByteArray(charset("UTF-8"))
        len = b.size
        if (len > ivBytes.size) len = ivBytes.size
        System.arraycopy(b, 0, ivBytes, 0, len)
        val keySpec = SecretKeySpec(keyBytes, "AES")
        val ivSpec = IvParameterSpec(ivBytes)
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec)
        val results: ByteArray = cipher.doFinal(text?.toByteArray(charset("UTF-8")))
        val EncryptedString: String =  Base64.encodeToString(results,0)
        return (EncryptedString.substring(0, EncryptedString.length / 2)
                + key
                + EncryptedString.substring(EncryptedString.length / 2))
    }

    fun generateRandomPassword(len: Int?): String {
        val chars = ("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijk"
                + "lmnopqrstuvwxyz+/")
        val rnd = Random()
        val sb = StringBuilder(len!!)
        for (i in 0 until len) sb.append(chars[rnd.nextInt(chars.length)])
        return sb.toString()
    }
}